### ISSUE TITLE
EE9 Compatible Release

### ISSUE THREAD

**Issue by codylerum** [ocpsoft/rewrite#304]
Are there plans for an EE9 compatible release which targets the jakarta namespace?

**Comment by lincolnthree** [comment]
Hey @codylerum ! Nice to hear from you :D It's been a while. 

So, I haven't had a chance to actually look at anything in EE9, and I'd love to support this but am not really sure what it would entail. Any details you can provide?

How complex would the migration be, and would it be backwards compatible? A PR would of course be welcome, but again, not sure of the complexity.

**Comment by lincolnthree** [comment]
Yikes. I wonder why they opted to do that. Maybe an easier way to deprecate and force replacement of ancient APIs.

**Comment by lincolnthree** [comment]
I'd be grateful if you wanted to take a stab at this!

**Comment by codylerum** [comment]
I believe it ended up being a copywrite issue when Oracle donated it to Eclipse. So there is going to be a firm demarcation between EE8 and EE9+

https://www.infoq.com/news/2019/05/end-of-javax-package/

Anyway I'll see if I can take a stab at it as I get closer to migrating my app to EE9. Honestly though I'm only using rewrite for a couple very simple things so I may just bring that in house and drop the rewrite dependency.

**Comment by BorisYellnikov** [comment]
So, I've changed POM and namespaces and I was able to compile Rewrite, excluding tests. After changing maven properties I had to implements 2 or 3 methods but I didn't know how to implements these methods :). So, it seems all the work is to implements 2 or 3 methods and changing POM to support Java 17 and EE 9. I think it would be not to much work to update Rewrite to EE 9.

**Comment by lincolnthree** [comment]
@BorisYellnikov Hey! Sorry, I missed this message. Would you mind sending a PR for what you've done so far? And let me know what is still incomplete / which methods are needing implementation? I'd love to make this happen, I just haven't had a lot of time to do the research to figure out where to start. (Since I'm not working with Java EE at the moment and my current project has been really all-consuming.)

I could most likely run with it and release a new version if you get me started! The tests are in sad shape :( Arquillian seems broken / unmaintained these days, unless I'm missing something.

**Comment by BorisYellnikov** [comment]
I did some time ago, so I don't remember exactly what i've done :). First what is to do is to change all namespace to jakarta. Next is to change in pom:
<dependency>
            <groupId>jakarta.platform</groupId>
            <artifactId>jakarta.jakartaee-api</artifactId>
            <version>9.0.0</version>
            <scope>provided</scope>
        </dependency>
         <dependency>
            <groupId>jakarta.servlet</groupId>
            <artifactId>jakarta.servlet-api</artifactId>
            <version>5.0.0</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>jakarta.inject</groupId>
            <artifactId>jakarta.inject-api</artifactId>
            <version>2.0.0</version>
        </dependency>

I don't remember if I had to change source code in rewrite or in my HttpConfigurationProvider. I remove also tests and run manven with -DskipTests=true. 

I've ran mavan today and this is result. I don't know why maven claims on dependencies.

NFO] Reactor Summary for rewrite-parent 3.4.5-SNAPSHOT:
[INFO]
[INFO] rewrite-parent ..................................... SUCCESS [  0.957 s]
[INFO] rewrite-test-base .................................. SUCCESS [  1.834 s]
[INFO] rewrite-api ........................................ SUCCESS [  2.622 s]
[INFO] rewrite-api-el ..................................... SUCCESS [  0.686 s]
[INFO] rewrite-impl ....................................... SUCCESS [  0.340 s]
[INFO] rewrite-addressbuilder ............................. SUCCESS [  0.771 s]
[INFO] rewrite-api-servlet ................................ SUCCESS [  1.071 s]
[INFO] rewrite-impl-servlet-2.5 ........................... SUCCESS [  0.603 s]
[INFO] rewrite-impl-servlet-3.0 ........................... SUCCESS [  0.527 s]
[INFO] rewrite-impl-servlet ............................... SUCCESS [  0.868 s]
[INFO] rewrite-test-harness ............................... SUCCESS [  2.031 s]
[INFO] rewrite-config-servlet ............................. SUCCESS [  2.254 s]
[INFO] rewrite-api-tests .................................. SUCCESS [  1.162 s]
[INFO] rewrite-impl-tests ................................. SUCCESS [  0.784 s]
[INFO] rewrite-annotations-api ............................ SUCCESS [  0.386 s]
[INFO] rewrite-annotations-impl ........................... SUCCESS [  1.301 s]
[INFO] rewrite-integration-cdi ............................ SUCCESS [  1.433 s]
[INFO] rewrite-config-annotations ......................... SUCCESS [  1.679 s]
[INFO] rewrite-config-jodatime ............................ SUCCESS [  0.905 s]
[INFO] rewrite-integration-faces .......................... SUCCESS [  1.838 s]
[INFO] rewrite-config-prettyfaces ......................... SUCCESS [  4.439 s]
[INFO] rewrite-config-prettyfaces-tests ................... SUCCESS [  1.777 s]
[INFO] rewrite-config-proxy ............................... SUCCESS [  0.760 s]
[INFO] rewrite-config-servlet-tests ....................... SUCCESS [  1.143 s]
[INFO] rewrite-impl-servlet-tests ......................... SUCCESS [  1.196 s]
[INFO] rewrite-transform .................................. SUCCESS [  1.097 s]
[INFO] rewrite-transform-less ............................. SUCCESS [  0.912 s]
[INFO] rewrite-transform-minify ........................... SUCCESS [  1.576 s]
[INFO] rewrite-transform-markup ........................... SUCCESS [  2.384 s]
[INFO] rewrite-integration-cdi-tests ...................... SUCCESS [  0.987 s]
[INFO] rewrite-security-integration-shiro ................. SUCCESS [  1.016 s]
[INFO] rewrite-servlet .................................... FAILURE [  0.390 s]
[INFO] rewrite-distribution ............................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  42.991 s
[INFO] Finished at: 2022-02-08T12:05:07+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project rewrite-servlet: Could not resolve dependencies for project org.ocpsoft.rewrite:rewrite-servlet:jar:3.4.5-SNAPSHOT: The following artifacts could not be resolved: org.ocpsoft.rewrite:rewrite-api:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-api-el:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-impl:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-config-servlet:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-api-servlet:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-impl-servlet:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-config-annotations:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-annotations-api:jar:sources:3.4.5-SNAPSHOT, org.ocpsoft.rewrite:rewrite-annotations-impl:jar:sources:3.4.5-SNAPSHOT: Could not find artifact org.ocpsoft.rewrite:rewrite-api:jar:sources:3.4.5-SNAPSHOT -> [Help 1]

**Comment by larsgrefer** [comment]
@lincolnthree Could you prepare a "develop"-branch for rewrite 4.0.0-SNAPSHOT which we can target our pull requests against?

**Comment by lincolnthree** [comment]
@BorisYellnikov Thanks! When you say "change all namespace to jakarta", can you give me an example?

That's an interesting error, I'm not sure what would cause the sources jars to be missing in the build. Or why it would fail on them. I'll look into it.

@larsgrefer Yes, I can do this tomorrow. Putting it on my todo list so I don't forget.

**Comment by BorisYellnikov** [comment]
This is old original namespaces
package org.ocpsoft.rewrite.servlet.http;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.ocpsoft.rewrite.event.Rewrite;
import org.ocpsoft.rewrite.servlet.ServletRewriteProvider;
import org.ocpsoft.rewrite.servlet.http.event.HttpServletRewrite;

/**
 * A {@link org.ocpsoft.rewrite.spi.RewriteProvider} that only operates on {@link HttpServletRequest} and
 * {@link HttpServletResponse} request cycle types.
 * 
 * @author <a href="mailto:lincolnbaxter@gmail.com">Lincoln Baxter, III</a>
 */
public abstract class HttpRewriteProvider extends ServletRewriteProvider<HttpServletRewrite>
{
   @Override
   public boolean handles(final Rewrite event)
   {
      return event instanceof HttpServletRewrite;
   }

   @Override
   public final void rewrite(Rewrite event)
   {
      if (event instanceof HttpServletRewrite)
         rewriteHttp((HttpServletRewrite) event);
   }

   /**
    * Handle the current {@link HttpServletRewrite} event.
    */
   public abstract void rewriteHttp(HttpServletRewrite event);
}

And new 

package org.ocpsoft.rewrite.servlet.http;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import org.ocpsoft.rewrite.event.Rewrite;
import org.ocpsoft.rewrite.servlet.ServletRewriteProvider;
import org.ocpsoft.rewrite.servlet.http.event.HttpServletRewrite;

/**
 * A {@link org.ocpsoft.rewrite.spi.RewriteProvider} that only operates on {@link HttpServletRequest} and
 * {@link HttpServletResponse} request cycle types.
 * 
 * @author <a href="mailto:lincolnbaxter@gmail.com">Lincoln Baxter, III</a>
 */
public abstract class HttpRewriteProvider extends ServletRewriteProvider<HttpServletRewrite>
{
   @Override
   public boolean handles(final Rewrite event)
   {
      return event instanceof HttpServletRewrite;
   }

   @Override
   public final void rewrite(Rewrite event)
   {
      if (event instanceof HttpServletRewrite)
         rewriteHttp((HttpServletRewrite) event);
   }

   /**
    * Handle the current {@link HttpServletRewrite} event.
    */
   public abstract void rewriteHttp(HttpServletRewrite event);
}

In general. All namespaces with javax should be change to jakarta in all EE specification namespaces. I think the most work will be to change all pom properties. Source code of rewrite should compile without problems without changes.

**Comment by lincolnthree** [comment]
@BorisYellnikov Ok thanks. This makes a lot of sense.

I created a #develop branch for this work.

**Comment by lincolnthree** [comment]
Started looking at dependencies and pom issues :sob:

**Comment by lincolnthree** [comment]
I'll fix the snapshot version later.

**Comment by larsgrefer** [comment]
@lincolnthree I've opened #310 as first attempt to make the project buildable again.
There are still some errors regarding the tests: `org.jboss.arquillian.container.spi.client.container.LifecycleException: Could not start container`

I'm not sure if its worth fixing the current containers, as we'll need JEE9 compatible ones anyways

**Comment by lincolnthree** [comment]
Thanks @larsgrefer - Merged. I agree about the containers. I also think deleting some of the *very* outdated integrations is probably in order. I think we can safely delete:

Apache Shiro
Google GWT
Tuckey URLrewrite config parser
Servlet 2.5
Servlet 3.0

Thoughts?

I got part way through the POM updates but had to stop. Planning on continuing over the weekend.

**Comment by lincolnthree** [comment]
PS. Does anyone know if there are any Arquillian updates for modern Jakarta EE containers? Is there a new equivalent test harness project we could use?

**Comment by larsgrefer** [comment]
@lincolnthree Servlet 2.5 and Servlet 3.0 can definitely go, since we'll have Servlet 5 baseline anyway.

With the minimum Java Version being 1.8 now, Joda Time should be redundant, too.

**Comment by lincolnthree** [comment]
Hey guys. Sorry, I got pulled into some stuff over the weekend. I have not forgotten about this. It might take me a bit but this is happening. My schedule is absolutely wild right now.

**Comment by larsgrefer** [comment]
> PS. Does anyone know if there are any Arquillian updates for modern Jakarta EE containers? Is there a new equivalent test harness project we could use?

At least Tomcat 10 seems to be supported: https://github.com/arquillian/arquillian-container-tomcat/pull/87

**Comment by lincolnthree** [comment]
@larsgrefer That's good to know. Merging your new PRs now. Still crazy over here. Thank you for continuing to work on this. I am really sorry I am not able to be more helpful just yet.

**Comment by lincolnthree** [comment]
@larsgrefer Wow, I am blown away by the incredible work you have put in to the past few PRs. Thank you, seriously. What maven command are you running the updated tests with? (E.g. container specification - if any - etc.)

**Comment by larsgrefer** [comment]
Just `./mvnw test` should run the tests.

In #316 I added a simple embedded Tomcat 9 to `test-harness` which fixes most of the Arquillian based tests.
I choose the embedded Tomcat because it was very easy to set up.

Some tests are still broken, but I'm not sure yet, why.
At least one test has a classloading issue which seems to be caused by the fact that the test itself and the Tomcat under test are running with different classloaders in the same JVM.

**Comment by larsgrefer** [comment]
@lincolnthree I think https://github.com/ocpsoft/common and https://github.com/ocpsoft/parent could use some updates to.

Are there any other ocpsoft projects which still use `common` and `parent` or can these be merged into the `rewrite` project?

**Comment by lincolnthree** [comment]
@larsgrefer A few projects use them. I can cut some releases if you want to send PRs to do what's needed! Probably best to increment the major version for these as well.

**Comment by larsgrefer** [comment]
`common` already got two PRs.

For new releases I agree with new major versions for each.
Since `common` uses `parent` we should release `parent` first, so `common` can use the new `parent` release

**Comment by lincolnthree** [comment]
@larsgrefer Parent and Common have been released. Please review my changes and let me know if you agree. I merged your very old PR on common (sorry about that,) and updated to Parent 16.

It might take a night for these to show up in Maven Central. But you should be able to manually `./mvnw clean install` from the Git tag in the mean time.

**Comment by lincolnthree** [comment]
Just checking in. Is there anything I can do to help at this point? I'm sorry again I haven't had a lot of time to focus on it, but i want to make sure I am not holding you up.

**Comment by larsgrefer** [comment]
You could have a look at the failing tests. https://github.com/ocpsoft/rewrite/issues/297 indicates that some tests are broken on master, too. It would be interesting to know, which tests are failing on both master and develop and which tests are only broken in develop.

You could also update the git branches in line with https://github.com/ocpsoft/rewrite/pull/321#issuecomment-1070970671
(Master could become 3.x or 3.5.x, develop could become 4.x or 4.0.x and 5.x/5.0.x could be a new branch)
There also seem to be some very old branches which could be removed.

**Comment by pdurbin** [comment]
@larsgrefer @lincolnthree hi! Any news? Anything I can do to help? I'm trying to upgrade our app (Dataverse) from Jakarta EE 8 and I can't deploy because we use rewrite-servlet and rewrite-config-prettyfaces. Removing them temporarily allows our app to deploy (I'm currently testing with Payara Server Community 6.2022.1 Alpha 2) but since we're using rewrite we either need an EE 9+ compatible version of rewrite or some other solution. Again, happy to help! Thanks for rewrite! Let me know!

**Comment by rtreacy** [comment]
It looks like I have something working for Dataverse/Payara6. Had to skip config-tuckey. Changes were like @BorisYellnikov described above. Off of develop branch (producing  *-3.60-SNAPHOT.jar files). Rush job when I found out from @pdurbin that it was going to hold up going to Payara6 Jakarta EE 9+

**Comment by lincolnthree** [comment]
Hi all, sorry I have not been able to keep track of this issue. Life has kept me unfortunately occupied.

@pdurbin @rtreacy You mentioned you got something working? Is that with or without Rewrite?

If you have any changes to submit back to this effort they would be welcome.

**Comment by pdurbin** [comment]
@lincolnthree hi! Yes! We got something working! Or rather, @rtreacy gave me a zip with changes he made and I pushed them (with slight modifications) into a fork of Rewrite at https://github.com/pdurbin/rewrite/commit/d494160

(From there I built the jars we need and added them in https://github.com/IQSS/dataverse/commit/d37af1d. Without these changes we can't deploy our war file to Payara 6 which is EE 9+ only. Our Rewrite/PrettyFaces rules seem to work fine.)

In that commit for Rewrite you'll see that I/we disabled `config-tuckey` and `distribution` and that we're skipping unit tests. We were just trying to get something working. 😄 

Where should we go from here? Would you like a proper pull request? The bulk of the changes are switching the namespace from `javax` to `jakarta`. If it's easier, please feel free to cherry-pick the commit and clean it up or do the equivalent.

We're in a bit of a time crunch because we use the "community edition" of Payara and as soon as version 6 comes out (EE 10 only and due at the end of the month) version 5 won't get security updates: https://www.payara.fish/what-payara-6-means-for-you/ . Worst case, we'll use those jars I built but I'd love to see the changes make it upstream into Rewrite.

**Comment by lincolnthree** [comment]
Hey @pdurbin, that's good to hear! Thank you for your effort, and for participating here to get this merged upstream.

Yes, I agree it is time to stop supporting Tuckey configuration.I think the best way forward right now is a pull request to the `develop` branch of the project, which is currently tracking this effort. Please feel free to send it, and I'll compare notes.

Once we get a working / releasable version, I will switch the branches and make develop 'main', and cut a release.

I totally understand your time crunch. I am in one myself.

**Comment by lincolnthree** [comment]
PS. Payara looks really interesting. What do you guys use to automate your testing? Anything we could use here in place of Arquillian? Seems like that project might be a bit obsolete at this point. At least not very active/updated.

**Comment by pdurbin** [comment]
@lincolnthree sure! I just created this PR:

- #322.

One thing I'm concerned about is that while we're testing with EE 9.1 right now soon we'll be testing with EE 10 (once Payara 6 Alpha 3 is out). I wonder if this means we'll have to bump `9.1.0` for Jakarta stuff in Rewrite in a couple places. At least the PR above has the namespace changes. I know this issue is for EE 9 compatibility but I think what we'll really need is EE 10 compatibility. I'm not sure what that means from the perspective of Rewrite. Maybe it's easy to support both. Or maybe this is a non-issue. I don't know.

Again, @rtreacy is the one who did the work so he'd know about the Tuckey stuff. I'm just the messenger and PR maker. 😄 

For automated testing we use REST Assured. I'd say we're pretty happy about it. You can read about how we use it at https://guides.dataverse.org/en/5.10.1/developers/testing.html

Thanks for taking a look at that PR!

**Comment by larsgrefer** [comment]
> @larsgrefer Is your branch currently functional (tests aside?)

My branch is essential your "develop" branch.

The big question is, if we (you) want to follow this plan here https://github.com/ocpsoft/rewrite/pull/321#issuecomment-1068237656 with 2 versions (one javax based (EE 8) and one jakarta based(EE 9+)) or if you want to jump to Jakarta EE 9+ directly

**Comment by lincolnthree** [comment]
@pdurbin So our `develop` is currently targeted at EE8, right? And your PR is based on EE9? If so, I will create a separate branch to merge that into so we can continue work in parallel.

I am inclined to agree with moving forward with two branches, though I suspect most work will be done going forward on EE9+, as `javax` truly is deprecated and has been for some time.

For the Tuckey stuff, I am not really interested in bringing that forward. I don't think it was really used much, and was generally intended to be a migration path for folks who wanted to stop using it. We can also nix the obsolete servlet API support.

I am particularly concerned with finding a path forward with testing. Figuring out what is failing, what matters, and what does not. Any help with that would be greatly appreciated.

**Comment by lincolnthree** [comment]
FYI. I went ahead and created a new branch [#develop-jakartaee-9](https://github.com/ocpsoft/rewrite/tree/develop-jakartaee-9) - and cleaned up all the other obsolete branches that were confusing things. I also renamed `master` to `main`.

So right now we have:
# main
# develop (ee8) 
# develop-jakartaee-9

**Comment by pdurbin** [comment]
Sounds good to me. With EE 10 around the corner, do we need a fourth branch?

- develop-jakartaee-10

I say this because Payara 6 Community Edition will only support EE 10.

**Comment by codylerum** [comment]
Is there any difference after the rename to the jakarta namespace? I would think anything on 9 would work with 10

**Comment by lincolnthree** [comment]
@codylerum That would be my assumption as well. 
@pdurbin Let's target EE9 as a 'minimum' version, then address any issues we need to once 10 comes out.

**Comment by lincolnthree** [comment]
Hey @pdurbin - Thanks for all the PRs. Mind talking me through your process on those? I just want to make sure I understand your intent!

Also, I saw you updated to Tomcat 10. Did you have any luck getting the tests to run on it?

**Comment by lincolnthree** [comment]
@pdurbin Did you see my questions on your PR? Just checking :)

**Comment by larsgrefer** [comment]
#324 is a work in progress for the transition to the jakarta.* apis. This basically began as a big search-and-replace `javax.*`-> `jakarta.*`. This was followed by some cleanup to make it compile and "work" again. In this process, I had to remove config-tuckey (because tuckey references `javax.` classes internally). I also updated to Tomcat 10 here to get some of the tests running again. (Deploying `jakarta.servlet`-based code to older Tomcats just gives ClassNotFoundExceptions)

I've then spilt the removal of config-tuckey out into 2 separate PR's for `develop` and `develop-jakartaee-9` respectively:
#327 and #328.

#325 and #326 remove joda time from `develop` and `develop-jakartaee-9` respectively, as its superseded by `java.time`

To have a better overview of which tests are failing where I opened #329 to run the tests on the main branch using GitHub Actions. I'm hoping to see, which tests are already failing on main, and which tests broke between `main` and `develop`.
Ideally when should fix (or remove) the failing tests on main, and merge this fixes into `develop` and `develop-jakartaee-9`

**Comment by pdurbin** [comment]
> @pdurbin Did you see my questions on your PR? Just checking :)

I just looked at #322 again and I don't see any questions. Happy to try to answer!

**Comment by larsgrefer** [comment]
some of the test failures seem to be caused by a version incompatibility between htmlunit and Apache httpclient.

On main, we use a "too new" httpclient version for the old htmlunit, that's why the faces tests fail.
If we update htmlunit to a newer version which works with the newer httpclient, the servlet tests fail, because htmlunit is now too new for selenium-htmlunit-driver

**Comment by lincolnthree** [comment]
@pdurbin Thanks, Hm. I wonder if reviews don't show up to all members / viewers?

![image](https://user-images.githubusercontent.com/362329/171656652-91ead015-bca3-43ff-959d-ec99c3ade9a8.png)

**Comment by lincolnthree** [comment]
@larsgrefer Thank you. This helps. So as I understand it, the Github actions branch should be merged into main - I tend to agree. Let's go ahead and do that. It will be good to get a baseline.

Would you say it is safe to merge the other PRs into the respective branches? Or would you like me to wait until we get a test baseline?

**Comment by larsgrefer** [comment]
> Would you say it is safe to merge the other PRs into the respective branches? Or would you like me to wait until we get a test baseline?

The ones not marked as draft should be safe to merge.
Since these PR's only remove things I'm not overly concerned about the tests (less features -> less tests -> less failing tests)

Ideally we would have a green build with no failing tests on main, which we could then merge upwards into the two develop branches. Then we would see whats _really_ broken. At the moment I see multiple problems with the tests:
- Some tests are already broken on main (No wonder they are broken on develop, too)
- Some failures might only result in the switch from Glassfish/Wildfly to Tomcat or in the switch from a "mananged" to an "embedded" container.
- The arquillian project seems to be quite inactive. I've not found that many resources on how to use more recent versions of Glassfish/Wildfly, etc.

On the one hand the extensive test suites are holding back the migrations (thats why @pdurbin just skipped all the tests) but on the other hand the tests are a valuable asset to check if we did everything correct. It would definitely be faster and easier, to just drop all the tests (and therefore arquillian), but I'm not sure if this would be a good idea long term.

**Comment by lincolnthree** [comment]
@larsgrefer Thanks. I'll move forward with merging the non-draft PRs.

I agree about the tests. Arquillian is a bit problematic here. I think the core functionality of the Rewrite framework is sound, and the critical features haven't changed since Servlet 2.5. They are very well tested, inherently. Most of the test failures, aside from the HTMLUnit ones, are likely fringe / edge cases that are container specific -- and likely fixes, from my experience, involve server configuration changes -- rarely framework / code changes..

 Worst case scenario, we do a release with tests skipped and wait for bug reports while we try to get the tests updated & back on line. 

Thoughts?

**Comment by codylerum** [comment]
I think that is fine.

What issues are we seeing with Arquillian? I know the wildfly guys are using it to test the new EE10 version of Wildfly and are getting closer to some non-preview builds of it.

**Comment by lincolnthree** [comment]
@codylerum I think most of the issues are related to package versioning. There have also been some deployment failures on newer containers and I haven't figured out how to resolve them. I don't have any stack traces at the moment, but each EE9/10 container seems to have its own quirks that block the tests from deploying.

I think on GlassFish the tests deploy once then fail to undeploy, and subsequent errors ensue. On WildFly I don't think I was even able to get a deployment at all.

**Comment by lincolnthree** [comment]
@codylerum I have an issue to track these problems started here: https://github.com/ocpsoft/rewrite/issues/297 (no more details there unfortunately. I was in a rush when I opened it.)

**Comment by codylerum** [comment]
@lincolnthree Sounds good. It looks like we should have something stable here in the next couple months and I'll try to chip in once i get my own projects working with Arquillian + Wildfly and EE10

**Comment by lincolnthree** [comment]
@larsgrefer Since you've been removing features. Want to take a stab at this one? 

https://github.com/ocpsoft/rewrite/issues/240

I think it's time to drop support for the older servlet specs with the update to EE9/10. (At least any update where the namespace has changed.)

**Comment by lincolnthree** [comment]
> @lincolnthree Sounds good. It looks like we should have something stable here in the next couple months and I'll try to chip in once i get my own projects working with Arquillian + Wildfly and EE10

Thanks @codylerum ! That would be amazing. PS. It's nice to "see" you again. I've been in a very distant world for a while and it's nice to see some familiar faces!

**Comment by larsgrefer** [comment]
> @larsgrefer Thanks. I'll move forward with merging the non-draft PRs.
> 
> I agree about the tests. Arquillian is a bit problematic here. I think the core functionality of the Rewrite framework is sound, and the critical features haven't changed since Servlet 2.5. They are very well tested, inherently. Most of the test failures, aside from the HTMLUnit ones, are likely fringe / edge cases that are container specific -- and likely fixes, from my experience, involve server configuration changes -- rarely framework / code changes..
> 
> Worst case scenario, we do a release with tests skipped and wait for bug reports while we try to get the tests updated & back on line. (E.g. Deleting the tests seems unnecessary.)
> 
> Thoughts?

I agree. We did close to none functional changes to the rewrite codebase. From what I've seen until now, all test failures are "infrastructure" problems caused by the switch from Wildfly/Glassfish to Tomcat and from switching from a "managed" arquillian container to a "embedded" one.

I think keeping - but ignoring/skipping - the tests for now is the best we can do.

**Comment by lincolnthree** [comment]
@larsgrefer Oops, sorry. I am not thinking right today :) Need more coffee! Thanks for confirming.

**Comment by larsgrefer** [comment]
I think i got most of the tests working again:

in #331 I divided the test into "normal" unit tests, which can run without arquillian and "integration tests" which will use arquillian. By doing so, I could focus on fixing the normal tests first.

in #330 I switched back from the embedded tomcat to Wildfly managed and wildly remote, which got most of the arquillian based tests working again. For the few tests which are still failing, I set `<testFailureIgnore>true</testFailureIgnore>`

**Comment by lincolnthree** [comment]
@larsgrefer That's awesome! So WildFly remote is working, but none of the other containers? This requires downloading and installing WildFly, as I believe Arquillian doesn't do that for REMOTE containers, right? Or did you automate that, too?

I went ahead and merged #331 - as #330 is still marked as draft, I didn't merge that one.

**Comment by larsgrefer** [comment]
The unit-test are executed by the maven-surefire-plugin in maven's `test` lifecycle-phase:
`./mvnw test`

For the integration test, we need maven's `verify` phase. I also re-introduced 2 maven profiles for the different ways of running Wildfly:
`./mvnw verify -P WILDFLY_REMOTE`
`./mvnw verify -P WILDFLY_MANAGED`

**Comment by larsgrefer** [comment]
@lincolnthree We should now decide which version numbers to use for which branch/version so we can start publishing snapshots.

The current status-quo is as follows:

| Branch  | EE Version | Rewrite Version |
| ------------- | ------------- | -|
| main  | Java EE 7  | 3.5.2-SNAPSHOT |
| develop  | Jakarta EE 8  | 3.6.0-SNAPSHOT |
| develop-jakartaee-9  | Jakarta EE 8  | 3.6.0-SNAPSHOT |

My suggestion would be this:

| Branch  | EE Version | Rewrite Version |
| ------------- | ------------- | -|
| main  | Java EE 7  | 3.5.2-SNAPSHOT |
| develop  | Jakarta EE 8  | **4.0.0**-SNAPSHOT |
| develop-jakartaee-9  | Jakarta EE 8  | **5.0.0**-SNAPSHOT |

**Comment by lincolnthree** [comment]
@larsgrefer I completely agree. I think this makes total sense. I'll work on updating versions today.

**Comment by larsgrefer** [comment]
Seems like the automatic snapshot deployments are now working: #332 https://github.com/ocpsoft/rewrite/runs/6864680645?check_suite_focus=true

So after you set the correct version numbers, we should forward-merge `main` into `develop` and then `develop` into `develop-jakartaee-9` in order publish snapshots for those versions, too.

**Comment by lincolnthree** [comment]
@larsgrefer Awesome! Version numbers are set. I had to force push and correct a simple mistake in the develop branch (within the span of about 2 minutes), so just in case you pulled during that time, that's what happened.

**Comment by larsgrefer** [comment]
@lincolnthree How would you like to continue with the two `develop` into `develop-jakartaee-9` branches?

Everything happening in `develop` should happen in `develop-jakartaee-9`, too, but we have 2 ways of doing this:

1. Doing everything twice: One PR for the change in `develop` and then a second PR for the same change in `develop-jakartaee-9`
2. Frequently forward-merging `develop` into `develop-jakartaee-9` in order to keep both branches in sync.

I think Option 2 is "cleaner" and more elegant, but the forward-merges should be done by someone with direct write access to the repository. Forward merges via PR (like #336) are a bit tedious

**Comment by lincolnthree** [comment]
@larsgrefer Sorry for the delay. Getting caught back up in firefighting work stuff again. My guess is that this will probably be mostly an issue doing bugfixes, because I think the framework is fairly feature complete, but yes, I tend to agree.

I think Option 2 is probably a bit cleaner, and I agree about a maintainer doing the work. This should be fairly painless assuming the two branches don't get too out of sync. 

I think Option 1 is fine if the contributor wants to go through that effort. I guess I don't really have a strong preference.

**Comment by lincolnthree** [comment]
@larsgrefer Can you update me on where the current PRs stand? There are a few open and I don't want to screw it up :)

Do you want me to merge the PRs targeted at `develop` so you can update your forward merge PR? Or do you want me to leave them alone for now?

**Comment by larsgrefer** [comment]
#337 and #339 would be up next

**Comment by lincolnthree** [comment]
@larsgrefer Merged. Those both look good. Thanks for splitting the integration tests. I honestly have not learned how GitHub actions work yet, so I really appreciate this.

**Comment by lincolnthree** [comment]
@larsgrefer You are incredible. Thank you for all of this work. I owe you many beers. What's next? All tests appear to be green. Is it time to cut a preliminary release?

**Comment by larsgrefer** [comment]
Releasing a 4.0.0-m1 and 5.0.0-m1 sounds like a good idea

**Comment by lincolnthree** [comment]
4.0.0.Alpha1 and 5.0.0.Alpha1 have been released in Sonatype Nexus and should sync to central by tomorrow at latest :) Thanks you @larsgrefer, @pdurbin, @codylerum, and @BorisYellnikov  for all of your amazing work in making this happen! And also thank you for your patience with me 😆 Apologies if I forgot anyone, I think @pdurbin was also submitting on behalf of a colleague.

Please let me know how it goes during testing.

**Comment by pdurbin** [comment]
@lincolnthree et al. thanks for the release and for reaching out! @poikilotherm has been digging in. Unfortunately, we're pretty sure even 5.0.0.Alpha1 isn't compatible with EE 10 so we opened #345. Please take a look when you get a chance. Thanks!

**Comment by lincolnthree** [comment]
Has anyone had a chance to test the Alpha releases for EE8 and 9 support? What was the outcome?

**Comment by pdurbin** [comment]
Our next target is EE 10 so we're really only testing that. That is, we're looking at #345. We don't plan to use EE 9. Thanks.

**Comment by rzo1** [comment]
> Has anyone had a chance to test the Alpha releases for EE8 and 9 support? What was the outcome?

I tested `5.0.0.Alpha1` on our jakarta migration branch and it worked fine.

**Comment by raphisuter** [comment]
> Has anyone had a chance to test the Alpha releases for EE8 and 9 support? What was the outcome?

Hi @lincolnthree
I don't know if this is still relevant but we are migrating a big platform from java EE7 to jakarta EE 9.1 and we are using rewrite. We are using 5.0.0.Alpha1. Can keep you updated if you're intressted.

**Comment by raphisuter** [comment]
Hi @lincolnthree 
Sadly at this point we are far of the JavaEE Standard at this point and need to sort out a lot of dependencies it is challenging to keep the overview of all the things popping up.
At this point I start to try and compile client after client and sort out all issues occuring. As I'm no at the point of trying to compile the first view related module using rewrite I got into troubles because it's missing a spring-framework dependency.


`Failed to execute goal on project view-client: Could not resolve dependencies for project ch.qualitas:view-client:jar:0.0-local: Failed to collect dependencies at org.ocpsoft.rewrite:rewrite-servlet:jar:5.0.0.Alpha1: Failed to read artifact descriptor for org.ocpsoft.rewrite:rewrite-servlet:jar:5.0.0.Alpha1: Could not find artifact org.springframework:spring-framework-bom:pom:6.0.0-M4 in nexus (inhouse nexus url) -> [Help 1]`


I'm not sure if I have to include this dependency if we are not using the spring integration. In the module which fails to build we use following dependencies


    <dependency>
      <groupId>org.ocpsoft.rewrite</groupId>
      <artifactId>rewrite-servlet</artifactId>
    </dependency>
    <dependency>
      <groupId>org.ocpsoft.rewrite</groupId>
      <artifactId>rewrite-integration-faces</artifactId>
    </dependency>

Any advice?

**Comment by pdurbin** [comment]
@lincolnthree hi! I see chatter here about 5.0.0.Alpha1 but should I inquire about 6.0.0.Alpha1 over in #345?

**Comment by lincolnthree** [comment]
This is released as rewrite Version `9.0.0.Final`! Thanks everyone for their patience and encouragement. This has been a momentous and multi-year effort. Thanks to @pdurbin @codylerum @larsgrefer @poikilotherm and everyone else.

![image](https://github.com/ocpsoft/rewrite/assets/362329/16dbb5ee-3aa5-4031-8f5b-9e2860b0e1b4)

**Comment by lincolnthree** [comment]
Closing this as COMPLETED!
